### DISCUSSION TITLE
How to differentiate a Virtual Method that should be Augmented vs. Replaced?

### DISCUSSION THREAD

**Discussion by najak3d** [dotnet/csharplang#5866] [Q&A] [answered]
I write two main types of Virtual Methods in a base class (often abstract).

1. Ones meant to be optionally augmented (i.e. MUST or SHOULD be called by overrides)
2. Ones meant to be optionally Replaced, and should NOT be called by the overrides.

What is the correct/best way to differentiate between the two, aside from a comment.    And the comment-method is problematic, because the Summary is supposed to tell you "how to use the method" not "how to override it".

So this has always been a bit of conundrum that I've seen, dealing with this ambiguity.  I wish there were a keyword to put in the base definition like "protected virtual extended void MethodName()" to signify a method that should be called by the overrides.  And then if the override method calls or doesn't call the base method, the compiler will issue a Warning (or error?) - to prevent the mistake.

**Example:**
We just fixed a long-standing bug, where one certain branch of the code failed to call the Android MainActivity base class method...  Since the branch was rare, it rarely happened, but when it happened, it was catastrophic.   Just as the compiler warns you when branches omit setting an "out parameter", it would be nice if a virtual method intended to be called by the overrides was likewise enforced as a compiler warning, at minimum.

**Comment by CyrusNajmabadi** [comment, marked as answer]
>  it would be nice if a virtual method intended to be called by the overrides was likewise enforced as a compiler warning, at minimum.

This would be a space for analyzers for your domain-specific additional rules you would like to have.

That said, the way i generally handle this is to never make it necessary for the override to call into the base.  Indeed, anything that must run is designed to be impossible not to run.  The override only exists for the part that is ok to be overridden without the base happening.

In other words.  INstead of:

```
public virtual void Foo()
{
    This.MustRun();
    This.DoesNotHaveTo();
}
```

I instead write:

```
public void Foo()
{
    This.MustRun();
    OptionalFooPart();
}

protected virtual void OptionalFooPart()
{
    This.DoesNotHaveTo();
}
```

Now there's no problem.  The subclass can't mess this up, because calling into the base is not necessary.

**Reply by najak3d** [reply]
That's actually what I do as well.  But have notice that many professional/mainstream libs don't follow this pattern, such as the Xamarin.MainActivity -- as we discovered via having a branch not calling it.

I thought it might be a good idea for C# to actually encode a RULE for this, along with some modifiers to virtual to indicate if it must be called, or shouldn't be called.   So that if you mistreat it, then it gives you, at minimum a compiler warning, or error.

Even in the pattern that we use, if there is a 2nd derivation of the class (one that derives from the derives class, you'll be right back in this boat again, given that the first derivation will either have code that must be called, or shouldn't be called if replaced.

The main thing that C# doesn't cover is to help prevent the calling of a base class that shouldn't be called if overridden.

I'm never satisfied with the solution of "writing analyzers" because the main goal of coming to this site is to "encode safety" directly into the C# spec, so that it prevents bad behavior world-wide, rather than only in few select projects that employ the right analyzers.  The goal is to improve the C# language itself, so that is has a wide/universal impact on quality/clarity.

I think it's blatant that virtual methods should be able to specify their usage as 
1. "must be called" -- error if not called
2.  "should be called" - warning if not called (can suppress locally)
3.  "should NOT be called" - warning if called
4. "must NOT be called" - error if called,
5. <No Specifier>.    Doesn't care.  No warnings or checks on usage.

Modifier keywords might be:  "callAlways", "callUsually", "callOmit", "callNever"

So method would look like this:

public alwaysCall virtual void MethodName() {}


Just brainstorming, and mostly came here to find out if C# had already solved this problem.

**Reply by CyrusNajmabadi** [reply]
> Even in the pattern that we use, if there is a 2nd derivation of the class (one that derives from the derives class, you'll be right back in this boat again, given that the first derivation will either have code that must be called, or shouldn't be called if replaced.

This should not be an issue. The derived type can seal their override and follow the same pattern I outlined above.

**Reply by CyrusNajmabadi** [reply]
> along with some modifiers to virtual to indicate if it must be called, or shouldn't be called. So that if you mistreat it, then it gives you, at minimum a compiler warning, or error.

As mentioned above, this is already possible today through the use of an analyzer.

**Reply by CyrusNajmabadi** [reply]
> I'm never satisfied with the solution of "writing analyzers" because the main goal of coming to this site is to "encode safety" directly into the C# spec, so that it prevents bad behavior world-wide, 

There's no difference. Anyone who would use this new language feature could use an analyzer designed for this purpose. So they then get that "encoded safety" and they have the bad behavior prevented "world wide". 

> rather than only in few select projects that employ the right analyzers.

This is not how analyzers work. You would literally ship this analysis with your library. So the consumer gets this without having to do anything themselves.

**Reply by najak3d** [reply]
When the solution becomes "ship an analyzer with our DLL/assemblies to ensure this protection", then it becomes "just our solution", and not world-wide.   But when the C# language specifies new keywords to modify usage of virtual methods, and Visual Studio/Code enforce it -- now it becomes worldwide, and many many more will use it.

I'm not planning to be in the business of producing C# techniques/solutions that become used worldwide.   I'm here, because the scope of this website/group is to do exactly that.

**Reply by najak3d** [reply]
Do you see the worldwide practical benefit of being able to specify via keyword the expected usage of your virtual methods?   This is a worldwide issue, that affects every C# project that employs virtual methods.   Solutions such as you/I have done to workaround this deficiency could become old news.

The issue with our technique of creating a new virtual method is that it not only takes extra code, but also pollutes the Intellisense API of available options -- now you have extra method names to sift-through, all because C# doesn't provide a better way to ensure proper usage of your Virtual methods.   PLUS -- many projects do not employ this workaround (such as the Xamarin Android package, e.g. MainActivity).

**Reply by HaloFour** [reply]
> Do you see the worldwide practical benefit of being able to specify via keyword the expected usage of your virtual methods?

I'm not aware of any OOP language that has a feature anything like this, so no, it does not appear to be a widely applicable feature.

**Reply by CyrusNajmabadi** [reply]
> When the solution becomes "ship an analyzer with our DLL/assemblies to ensure this protection", then it becomes "just our solution", 

It's not though. It's something someone can build once and then libraries can use as much as they want.

**Reply by najak3d** [reply]
> > Do you see the worldwide practical benefit of being able to specify via keyword the expected usage of your virtual methods?
> 
> I'm not aware of any OOP language that has a feature anything like this, so no, it does not appear to be a widely applicable feature.

It takes imagination to be the first to solve a problem, or to make an improvement, that others have simply learned to work with.  For example, many do what Cyrus and I do - which is to create a 2nd do-nothing virtual method wrapped by a non-virtual method that calls it... so that it doesn't matter if the override calls it or not.   So this work around requires more code, and double the members on the API (more names to sift through) -- all to get around this C# deficiency.

So the problem is common enough, that the most experience C# programmers have derived workarounds like this, and have simply become "used to it".  So it's common - and the impact of an enhancing C# would be to deprecate these workarounds, and simply allow us to write one virtual method, where the usage is enforced.   This cuts down on code, cuts down class members, and may even cause more programmers to make use of it, than who already have adopted work around solutions.

Even the authors of Xamarin Android don't know about this workaround... so failing to call the base method, things fail, but in ways that you might not detect until much later.    If they had simply had a "must be called" keyword for their method, then we could not have made the mistake we made (and I know we're not alone here; it's a common concern/problem).

I think the key word should be a combination of 1 or 2 terms (2nd term is optional):

1st term:
1. mustCall / shouldCall  = error / warning if you don't call it
2. mustOmit / shouldOmit = error /warning if you DO call  (warning can be suppressed)

2nd term-- applies only for "mustCall / shouldCall":
1. First -- the call must come at the start of the override, or it gives warning (can be suppressed)
2. Last -- the call must come at the end of the override, or it gives warning (can be suppressed)

2nd term is only a warning -- so it is "informative/instructional" in nature, so that you know the intended usage.

Syntax would something like:

protected virtual mustCallFirst void _MyVirtualMethod() { }

Then overrides that fail to call it -- generate a compiler error.
And overrides that fail to call it First -- generate a suppressible warning

**Reply by HaloFour** [reply]
> It takes imagination to be the first to solve a problem, or to make an improvement, that others have simply learned to work with.

Or, more likely, it's not enough of an issue.  In my case it's not that I've ever had to work around this, the situation has simply never come up.

> For example, many do what Cyrus and I do - which is to create a 2nd do-nothing virtual method wrapped by a non-virtual method that calls it... so that it doesn't matter if the override calls it or not. So this work around requires more code, and double the members on the API (more names to sift through) -- all to get around this C# deficiency.

Languages don't optimize syntactically for the uncommon especially when other solutions exist.  There doesn't need to be a new keyword for every possible weird situation that will arise.  If in the rare case you need two building blocks instead of one that is more than sufficient.

> and I know we're not alone here; it's a common concern/problem).

If it were it probably would have come up before now, don't you think?

**Reply by CyrusNajmabadi** [reply]
> So the problem is common enough, that the most experience C# programmers have derived workarounds like this,

I don't see how it's a workaround.  It's like saying: oh man, i want to have a private field, but i want to expose a way to manipulate it, and now i have to have a method.  This is so common, and everyone has derived the workaround of having a public method to manipulate it.  So shouldn't we have a way to do this without the method?

The fact that something is easy and simple to do with existing language constructs does not lead to a motivation to then make all those easy things possible just through a new language unifying language construct.  Often it means: there is no problem here and the language is totally sufficient, it's just that you should be using the language as intended, not making a feature to make up for problems that arise when you do things you end up not liking later on.

**Reply by CyrusNajmabadi** [reply]
> It takes imagination to be the first to solve a problem

Maybe there's no problem here?   For example, if someone makes a field public and hten says: "man... i wish there were a way to make it so that people can't actually access it", then we don't say "we need a new access modifier!"  we will often say "why not use the access modifiers already available?"

I get the sense that this is a self-inflicted problem.  e.g. you've used the language in a way that has consequences, but you don't like those consequences.  And while the language has existing constructs without those consequences, you don't want to use them.  But that's not a motivation for us to make changes here.  

> Even the authors of Xamarin Android don't know about this workaround...

If they don't know about how to do this with the existing language, why do you think adding a new language construct would change things at all for them?

> I think the key word should be a combination of 1 or 2 terms (2nd term is optional):

ALl of this is possible with analyzers.  WHich is a first class supported mechanism to add limitations to the language after the fact.  That you ahve arbitrarily chosen not to use available mechanisms to achieve these goals is not a motivation for us to add new mechanisms instead.

**Reply by najak3d** [reply]
The problem here is that most people who write "protected virtual" methods aren't thinking about the ways in which it can be misused.  Since they are often both the author of the base class, and the derived classes, they simply use them "appropriately" according to their own knowledge.   If they forget to call the base class - they just say "oops, I should have known that", etc. 

Most people aren't thinking more critically about this stuff.  Mistakes are made, and that's just status quo.

It's very similar to the drive towards "records" and "immutability" -- most programmers don't naturally identify mutability of a class as a problem.   More experienced programmers also -- still might not recognize the problem, even as they are frustrated by the bugs that occur due to mutability.  And very few, probably realized that best solution was to make C# have better support for Immutability, to make it easier and more prevalent to be concerned for this.

Same things goes for the Nullable support recently added....   even though "null errors" plague code, and might be the #1 error that occurs, very few thought "C# should make this easier to avoid these error"...  they just say "oops, I forgot to null check" and move on.

Same thing with "protected virtual" methods -- most don't see the problem/risk of the current non-specific notation, for overrides that won't know how to appropriately call (or not call) the base method.  And there is no compiler support to catch these type mistakes.

IMO, there should be.  It should be mainstream, and part of the C# language to make this clear to the compiler -- so that mis-use will be considered a warning or error, and avoid the types of bugs that often don't get discovered until after release.

It's a serious enough issue, that @CyrusNajmabadi  and I have both derived a work around to ensure "proper usage" by removing the risk -- but MOST developers DO NOT DO THIS (when they probably should).    The work-around is inferior to simply having a couple new keywords, to dictate usage on a virtual method.   The work-around is: 

1. Not advertised, or well known
2. Doubles the number of members for each virtual method
3. Requires a fair amount of extra code (an added method).

**So would be nice if C# added some keywords on Virtual method to dictate usage in the overrides.**

**Reply by theunrepentantgeek** [reply]
You are assuming that the author of the base class can anticipate every possible way their class might be extended by _any_ developer _anywhere_ and _anywhen_.

I'm very sure I can't do that in the general case, and only infrequently in the specific case.

Extensibility of most classes is poorly designed, if it's designed at all. 

I literally encounter the problem of classes that are not extensible enough _orders of magnitude_ more often than class that are too extensible. 

There are some practitioners who genuinely believe that this is enough of an issue that _nothing_ should be **private**. Everything should be **internal** or **public**. While I disagree with the position (class invariants are too important), I have some sympathy for the position. 

Your original problem - failing to call the base method - sounds far more like a deficit of testing more than anything else. 

Whether you like it or not, this kind of special case is exactly why analysers were introduced c

**Reply by CyrusNajmabadi** [reply]
>  The work-around is:
>
> Not advertised, or well known
> Doubles the number of members for each virtual method
> Requires a fair amount of extra code (an added method).

Your second and third points are the same here. But as I said, this is not a negative in my book. That said, if it is a negative for you, there is an easy solution which has been available for years: analyzers. 

This sort of case is exactly what they were created for and what people already use them for. 

> So would be nice if C# added some keywords on Virtual method to dictate usage in the overrides.

We are not likely to do this when a completely suitable approach is instead to have an analyzer that adds she attributes to dictate usage in the overrides.

**Reply by najak3d** [reply]
@CyrusNajmabadi :  The issue of Immutability could have also been implemented fully with custom Attributes and analyzers.  You could have marked all readonly fields with "[Readonly]" and then use analyzers to determine violations.  So why should C# address immutability, since it can already be handled without C# upgrades?

Answer:  Because the C# enhancements made enforcement of immutability much easier, and saves code and bugs.   Same can be said for Nullability, as well as the majority of C# enhancements.

Your continued line of arguing would lead us to believe that C# 4.0 was fine, and didn't need to be changed.

And:
1. Doubles number of members 
2. Added extra code 

Are NOT the same Con.    Extra code doesn't always (or even usually) result in added class members.   The cost of extra code, costs you ONCE locally.  The cost of adding members costs the USERS of your class, as well as those maintaining your class, now seeing extra members, to decipher.  And introduces new opportunities for defect (derived classes calling your new method, when they should NEVER call it).   So these are TWO cons, not just one.

Shouldn't this have already been obvious to you?

**Reply by najak3d** [reply]
@theunrepentantgeek  - The fact that you can't control how others will override your virtuals is exactly why they might not make them virtual, AND they don't want to add more Members to the API, so they don't employ Cyrus's (and my) technique of adding additional members with specialized names.    It would be a heck of a lot nicer, if "virtuals" *could* have influence over the overrides (rules/guidelines for overrides).

Currently Virtual methods have a "summary" comment, but these summaries are about "functionality" and shouldn't be cluttered with "instructions for override" -- and so those overriding usually have no instructions.

Analyzers of this would also be crippled, because the analyzers ALSO do not have a clue as to usage rules for each virtual member, and so won't know when to issue Warnings.  **Only the author of the base class knows usage intent, but has no feasible facility to communicate this to derived classes.**

Also -- these suggested prescriptive keywords are OPTIONAL.   Just use them if you KNOW... plus I've suggested keywords to either trigger an Error vs. Warning -- depending upon whether the author of base class wants to communicate a "MUST" vs. a "SHOULD", vs. the absence of keyword means "I have no idea" (which is the current status for ALL protected methods now).

**I'd like to make C# better for the default/common case of opening up your class to overrides, with more QA control.**

**Reply by CyrusNajmabadi** [reply]
>  The issue of Immutability could have also been implemented fully with custom Attributes and analyzers.

Immutability is exactly something I would be ok with an analyzer/SG approach to.

**Comment by najak3d** [comment]
NEVERMIND -- just looked it up -- "sealed" is ALREADY a thing for preventing the propagation of overrides.

**IGNORE THIS QUESTION:**
Another question - can a derived class SEAL the virtual status of the method?  So that derivations of the derivation CANNOT override the derived override?

**Reply by FaustVX** [reply]
Yes with [`override sealed`](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQGYACLEgYRIG8CT6zSA3KAJwBcBXCAGzIBYSAWQAUAShp0GAXwKz8BYmUwkAIiRCVJ+BoxIB7JgFNWrKAiMkAzkd5GEA4eO2750oA===) modifier

**Comment by najak3d** [comment]
For case of making my concerns more clear.  Here is an example of work arounds that I've done in the past:

```c#
public void Update(float timeStep)
{
     // Core non-optional logic that must run goes here ....
    // .....

    OnUpdate();
} 

virtual protected void OnUpdate() {  }  // does NOTHING unless overridden
```

Here, those who write derived classes will see BOTH "Update" and "OnUpdate", and may call the wrong one.  We've introduced more ways to mess up, and more thinking for the dev to do -- for no good reason, other than a C# deficiency in allowing the base class author to specify usage rules.

What I'd prefer to write is this:
```c#
public virtual mustCallFirst void Update(float timeStep)
{
     // Core non-optional logic that must run goes here ....
    // .....
} 
```

This shorter code snippet produces the SAME effect, forcing the override to call the base method first, and requires no additional class member to enforce it.

This added syntax puts control on usage where it belongs - with the author of the base class, without forcing the addition of "similar named" members to create the same effect.

**Comment by najak3d** [comment]
I'd add to this proposal, specifiers that should even limit "who can call a method".    If for some reason I do still need to create a pattern like this:

(introducing combo term **"protected private"**)

```c#
public void Update(float timeStep)
{
      _OnBeforeUpdate();

     // Core non-optional logic that must run goes here ....
    // .....

    _OnAfterUpdate();
} 

virtual protected private void _OnBeforeUpdate() {  }  // does NOTHING unless overridden
virtual protected private void _OnAfterUpdate() {  }  // does NOTHING unless overridden
```

In the above example, I've introduced a suggested notation of **"protected private"** -- which means "you can ONLY OVERRIDE this method" -- you cannot CALL this method from a derived class (because it is ONLY intended to be called by the "Update()" method -- no one else.


This highlights another issue with the @CyrusNajmabadi  (and my) methodology of creating an extra member to help prevent "misuse" -- we are now opening up another potential defect, in that people can call the wrong method, when they should NEVER be allowed to CALL this other method (only override it).

**C# does NOT provide enough "control/prescription on usage" of it's protected methods, nor for it's virtual methods.  This is worth serious discussion.**

**Reply by HaloFour** [reply]
Constantly repeating something doesn't make it true nor does it do much to further the conversation.  Your points are heard, but thus far are not found compelling enough to consider changes to the language.  Certainly not to the degree of adding four new accessibility modifiers.  Analyzers were added to the compiler _very specifically_ to allow projects to add specific requirements like this.

**Reply by najak3d** [reply]
My point in bringing this up here, is that I don't think this is a niche need - it's universal.

I get that every good idea doesn't warrant a change to C# to accommodate.   But I keep arguing this case, because apparently it's not something that has been discussed (much) before -- so am trying to give the concept some exposure before cowering away and dropping it.

As with most "things of habit" -- after you've lived with something for long enough, you stop seeing any reason to change it.  I don't think most people have thought about the deficiencies of C#'s current inability to prescribe usage of it's "virtual" methods.   The improvements I am suggesting, would be a benefit, IMO.

**In short,** 
1. Just because you want to make behavior overridable - does NOT mean you want derived classes to CALL this override.  How to prevent that?   It's clear that "protected" says "you can call me", but "protected virtual..." is often only "protected" for sake of "override" NOT for sake of "calling it"!...   In C#, you cannot communicate this difference.    "protected private" would do this, closing the QA issue, as well as reducing API members seen in intellisense.

2. Once you do allow overrides -- you open yourself up to risks of misuse.   The ways to handle this in C# are non-elegant, and introduce additional ways to misuse or add defects.


I repeat myself because communication is hard.  Often it takes repetition for a new idea to sink in.

**Reply by najak3d** [reply]
If just \# 1 above were implemented, i.e. "protected private", that would eliminate one of the chief Cons of the current workaround (i.e. creating more methods).   At least then, your new methods would not create more API for derived classes.... they would only see this method show up in Intellisense as an option for "override" -- never for calling it.

That seems like an EASY change to C#, no?  And provides excellent benefit.  I'd think something like this is low-hanging fruit.

**Reply by najak3d** [reply]
After this helpful discussion, I think I'm going to raise # 1 above to it's own "Idea" thread.   As it seems like less-of-a-stretch than the concept of "mustCallFirst" (etc) key words.

**Reply by HaloFour** [reply]
> My point in bringing this up here, is that I don't think this is a niche need - it's universal.

It's universally applicable, sure, but it's a niche use case.

> Once you do allow overrides -- you open yourself up to risks of misuse. The ways to handle this in C# are non-elegant, and introduce additional ways to misuse or add defects.

By that thinking _every_ OOP language is inherently defective, yet this is not a common problem, not in C# or any other language.  You guide your consumers by the shape of your API, and if that shape necessitates multiple methods then so be it.  The language isn't going to add completely new features for uncommon scenarios just to make it slightly easier than all of the existing solutions that have existed for many, many decades.

> If just # 1 above were implemented, i.e. "protected private"

`private protected` is already a legal modifier in C#.

**Reply by najak3d** [reply]
Thanks @HaloFour  for the info on "private protected".   I've altered my suggestion to "private virtual" instead -- to mean "Overridable but NOT callable", which IMO, makes sense for that to mean this.

**Reply by CyrusNajmabadi** [reply]
> My point in bringing this up here, is that I don't think this is a niche need - it's universal.

This is a very common idea on a lot of your posts. Even reinforced on your part by stating your surprise that other people are not asking for this (despite thousands of requests asking for other things). You then follow things up with justifications that people don't even realize they're missing this and it truly is necessary. 

I truly ask that you examine your biases here.  We get tons of feedback about an amazing wealth of topics. Consider truly that if only you are vehemently asking for something that perhaps it really is not universal, and that not everyone else is ignorant and doesn't even realize they're missing this clearly obvious feature that you feel provides so much value.

**Comment by najak3d** [comment]
I added a new Language Idea here, for y'all to shoot down.   Enjoy.

https://github.com/dotnet/csharplang/discussions/5885

**Reply by CyrusNajmabadi** [reply]
> for y'all to shoot down.

@najak3d this style of discourse is not really appropriate here. Please be respectful of others here and accept that that may genuinely disagree on merit and not for the sake of disagreement.
